server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/StandaloneServerLocalConfigurationStorage.java

```diff
@@ -0,0 +1,142 @@
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Map;
+
+import org.infinispan.commons.CacheConfigurationException;
+import org.infinispan.commons.api.CacheContainerAdmin;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.globalstate.LocalConfigurationStorage;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.client.ModelControllerClient;
+import org.jboss.dmr.ModelNode;
+
+/**
+ * A {@link LocalConfigurationStorage} which saves {@link org.infinispan.commons.api.CacheContainerAdmin.AdminFlag#PERMANENT}
+ * changes to the server configuration model in standalone mode.
+ *
+ * @author Tristan Tarrant
+ * @since 9.2
+ */
+
+public class StandaloneServerLocalConfigurationStorage implements ServerLocalConfigurationStorage {
+   private static final String[] CACHE_MODES = {"local-cache", "invalidation-cache", "replicated-cache", "distributed-cache", "scattered-cache"};
+   private ModelControllerClient modelControllerClient;
+   PathAddress rootPath;
+   private EmbeddedCacheManager embeddedCacheManager;
+
+   @Override
+   public void initialize(EmbeddedCacheManager embeddedCacheManager) {
+      this.embeddedCacheManager = embeddedCacheManager;
+   }
+
+   @Override
+   public void validateFlags(EnumSet<CacheContainerAdmin.AdminFlag> flags) {
+   }
+
+   @Override
+   public void createCache(String name, String template, Configuration configuration, EnumSet<CacheContainerAdmin.AdminFlag> flags) {
+      if (!flags.contains(CacheContainerAdmin.AdminFlag.PERMANENT)) {
+         throw new CacheConfigurationException("Cannot create a non-permanent cache");
+      }
+
+      if (template == null) {
+         throw new CacheConfigurationException("Cannot create a cache not based on an existing template");
+      }
+
+      String cacheMode = findCacheMode(template, true);
+
+      if (cacheMode == null) {
+         throw new CacheConfigurationException("No such template " + template);
+      }
+
+      final ModelNode cacheAddOp = new ModelNode();
+      cacheAddOp.get(OP).set(ADD);
+      cacheAddOp.get(OP_ADDR).set(rootPath.append(cacheMode, name).toModelNode());
+      cacheAddOp.get("configuration").set(template);
+
+      try {
+         ModelNode resp = modelControllerClient.execute(cacheAddOp);
+         if (!SUCCESS.equals(resp.get(OUTCOME).asString())) {
+            throw new CacheConfigurationException("Could not create cache");
+         }
+      } catch (IOException e) {
+         throw new CacheConfigurationException("Could not create cache", e);
+      }
+   }
+
+   @Override
+   public void removeCache(String name, EnumSet<CacheContainerAdmin.AdminFlag> flags) {
+      String cacheMode = findCacheMode(name, false);
+
+      if (cacheMode == null) {
+         throw new CacheConfigurationException("No such cache");
```

Comment:
"No Such ..." doesn't sound right to me (although I think we have similar messages elsewhere), `"Cache '%s' does not exist", name` is more informative and clear IMO. The same also applies in `createCache(..)`.